Feature/drafts

cheapscate7 opened the pull request:
implements #18 

Adds "drafts" as a "pre-step" to the current resource workflow, ie, only non-submitted resources can be drafted.

Adds "delete by ID" tool in admin resource approval to help with removing abandoned drafts

aberonni commented on prisma/schema.prisma:
```diff
@@ -41,6 +41,7 @@ model Resource {
   type          ResourceType
   categories    CategoriesOnResources[]
   configuration ResourceConfiguration
+  draft         Boolean                 @default(false)
```
I suppose defaulting this to `true` would make prisma complain about updating existing records, while the default of `false` isn't quite a good reflection of reality.

I would actually love an alternative along the lines of

```diff
+ enum ResourcePublicationStatus {
+   DRAFT
+   READY_FOR_REVIEW
+   PUBLISHED
+ }


...
- published         Boolean  @default(false)
+ publicationStatus ResourcePublicationStatus                 @default(DRAFT)
```

because I think this would simplify the code in a lot of places, and also TS would force us to make sure that we're making the appropriate changes where "published" was being referenced before.

aberonni commented on prisma/schema.prisma:
```diff
@@ -164,17 +165,17 @@ enum UserRole {
-  accounts      Account[]
-  sessions      Session[]
-  resources     Resource[]  @relation("Resources")
+  id              String       @id @default(cuid())
```
I'm wondering where these changes came from? I'm surprised to see them

cheapscate7 replied:
Ooh you are right, i didnt expect to see them either. honestly, I couldnt quite tell you

cheapscate7 replied:
thanks for the tip

aberonni commented on src/pages/admin/pending-publication.tsx:
```diff
@@ -18,6 +61,12 @@ export default function MyContributions() {
         authenticatedOnly={[UserRole.ADMIN]}
       >
         <ResourceList queryResult={queryResult} useFilters showEditProposals />
+        <div className="my-8 lg:max-w-sm">
```
While this is a good idea in theory, I don't think it's worth the additional complexity in the codebase because the same thing can be easily done directly from prisma studio.

cheapscate7 replied:
Thats fair, I guess I felt a little guilty about not having any way to manage drafts as an admin within the app. Ive removed it

aberonni commented on src/pages/resource/[id].tsx:
```diff
@@ -121,13 +121,11 @@ export const SingleResourcePage: NextPage<{ id: string }> = ({ id }) => {
-  }
+  if (!resource) return <div>404</div>;
+
+  if (resource.draft && resource.createdById !== user?.id) return <div>403 - Forbidden</div>
```
This kind of check should happen only in the API (because otherwise we're giving users the possibility to still download the resource but just not showing it to them in the UI)

cheapscate7 replied:
understood, i think the resource api already covers this scenario anyway

aberonni commented on src/pages/user/my-proposed-resources.tsx:
```diff
@@ -1,26 +1,58 @@
 import { api } from "@/utils/api";
+import { Button } from "@/components/ui/button";
+
+enum TABS {
```
With the proposed `publicationStatus` flag, then we wouldn't need the tabs, and we would just show a different badge on each row.

cheapscate7 replied:
Personally, I would like to be able to have them separate, at least on this page. I will change it if you really dont like tabs though

cheapscate7 replied:
Thats fine, Im not that attached to the tabs

aberonni commented on src/server/api/routers/resource.ts:
```diff
@@ -344,7 +366,7 @@ export const resourceRouter = createTRPCRouter({
       };
     }),
-  update: adminProcedure.input(resourceCreateSchema).mutation(updateResource),
+  update: privateProcedure.input(resourceCreateSchema).mutation(updateResource),
```
This change is a risky one. We need to add an extra safe-guard in this method to make sure that the user can edit the resource they are trying to edit.

cheapscate7 replied:
You are right, Ive pushed a commit that adds an `OwnerProcedure` for mutations like this that should be a step up from `privateProcedure`

aberonni reviewed (commented):
Thanks for the contribution! This is really cool. 

I've added a few comments here and there, but after seeing it in practice I've actually thought up a process that I think I would prefer to this one. I'm gonna outline it here, but you don't have to implement it if you're not up for it. I'm happy to run with it based on your proposal. Lemme know what you think.

- When a user hits "propose new resource", they first see a dialog where they are prompted to enter title and URL
- After submitting the title and URL, we create the resource in the backend (with `publicationStatus: DRAFT`), and redirect the user to the "edit" form (so the "create" page does not exist anymore)
- In the "edit" form
  - We leave the "Save" button but we also add a "Submit for review" button
  - We update the alert to explain that you can save as much as you like before submitting for review. Once you hit "submit for review", you can't edit or save anymore
  - When the user hits "Submit for review", we simply switch to `publicationStatus: READY_FOR_REVIEW`  

This would also open up the possibility in the future to fetch similar resources from the backend after the first step (the title dialog) and say "Are you maybe trying to re-create an existing resource?" and show the alternative resources. I wouldn't do this in the first run, but it's an idea.

I think this would make for a smoother, easier to understand UX, and also reduce the amount of complexity we need to add in the codebase. What do you think?

cheapscate7 commented:
@aberonni I perfer your proposal workflow. My Idea originally was just for it to not touch too much of what was already there (hence being a "pre"-step) but your way is much more efficient. I didn't want to respond until I'd had the chance to try it out but I ended up making an implementation in the process so let me know what you think

aberonni commented on src/components/are-you-sure-dialog.tsx:
```diff
@@ -1,4 +1,4 @@
-import { useEffect, useMemo, useState } from "react";
+import React, { useEffect, useState } from "react";
```
If I'm not mistaken, all the changes in this file can be reverted because the new dialog isn't being used anymore.

aberonni reviewed (changes requested):
I didn't dig deep into the implementation because it seems that things still need to be addressed (eg. CI failures, and still some dead code from the previous implementation) so I assume this is still a WIP. That said, I love the new direction and would love to see it through to completion!

I imagine it would be easier for you to test with a real DB. If you hit me up (eg. on LinkedIn) I'd be happy to give you access to a development db with a copy of production data so you can test this more easily, and so that we can get a migration in place on planetscale before eventually merging.

cheapscate7 commented on src/pages/resource/[id]/edit.tsx:
```diff
@@ -81,9 +69,8 @@ export const ResourceEditPage: NextPage<{ id: string }> = ({ id }) => {
-              return;
-            }
+            if (isSubmitting) return;
+            //@ts-ignore
```
There was some sort of error, cant remember what but it made no sense with the types provided. Also, upon testing, it didnt seem to function unexpectedly. Probably was the Typescript language server as ive removed it to see what the error was and there isnt any anymore.
Long and short, its gone now

cheapscate7 commented on src/components/resource-create-form.tsx:
```diff
@@ -0,0 +1,123 @@
+import React, { useEffect } from "react";
```
true

cheapscate7 commented:
Thanks! yes, I wanted to get your opinion first to make sure I understood what your suggestion was as I dont have a lot of time in the coming days. 
Ive already addressed some of your comments but I will have to get around to some of the other stuff like migrations In a bit

aberonni commented:
Yeah there's no rush at all so you can take your time. If you'd prefer me to take it on from here that's also fine. Just let me know either way. If you do decide to keep at it, ping me elsewhere so I can give you access to a dev DB branch.
